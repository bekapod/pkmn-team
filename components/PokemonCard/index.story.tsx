// tslint:disable-next-line:no-implicit-dependencies
import { storiesOf } from "@storybook/react";
import React from "react";
import PokemonCard from ".";

storiesOf("PokemonCard", module)
  .add("default", () => (
    <PokemonCard
      pokemon={{
        id: "25",
        name: "Pikachu",
        slug: "pikachu",
        pokedexId: 25,
        sprite: "25.png",
        types: [{ name: "Electric", slug: "electric" }]
      }}
    />
  ))
  .add("when has actions to display", () => (
    <PokemonCard
      pokemon={{
        id: "25",
        name: "Pikachu",
        slug: "pikachu",
        pokedexId: 25,
        sprite: "25.png",
        types: [{ name: "Electric", slug: "electric" }]
      }}
      renderCardActions={(): JSX.Element => (
        <div>Card actions that should be rendered</div>
      )}
    />
  ));
