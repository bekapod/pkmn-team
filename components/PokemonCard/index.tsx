import React, { ReactNode, ReactNodeArray } from "react";
import styled from "styled-components/macro";
import { formatPokemonName, sortTypes } from "../../helpers/general";
import * as variables from "../../helpers/variables";
import { Pokemon, Type } from "../../types"; // eslint-disable-line import/named
import { CardContent, CardHeader, CardWrapper } from "../Card";
import CardHeading from "../CardHeading";
import InlineList from "../InlineList";
import TypeTag from "../TypeTag";

interface Props {
  memberId?: string;
  pokemon: Pokemon;
  renderCardActions?: () => ReactNode | ReactNodeArray;
}

const PokemonCardContent = styled(CardContent)`
  padding: ${variables.spacing.lg}px 0;
  align-items: center;
`;

const PokemonCardSprite = styled.img`
  height: calc(${variables.spacing.lg}px * 3);
`;

const PokemonCardActions = styled.div`
  margin-top: ${variables.spacing.lg}px;
`;

const PokemonCard = ({
  memberId,
  pokemon,
  renderCardActions
}: Props): JSX.Element => {
  const { pokedexId, types, name, sprite } = pokemon;

  return (
    <CardWrapper data-testid={`pokemon-${pokemon.id}`}>
      <CardHeader types={types}>
        <CardHeading>{formatPokemonName(pokemon)}</CardHeading>
      </CardHeader>

      <PokemonCardContent>
        <PokemonCardSprite
          src={`/static/sprites/${sprite}`}
          alt={`${name} sprite`}
        />

        <InlineList>
          {sortTypes(types).map((type: Type) => (
            <li
              key={`${
                memberId ? `Member: ${memberId}` : ""
              } Pokemon: ${pokedexId}, Type: ${type.slug}`}
            >
              <TypeTag type={type.slug}>{type.name}</TypeTag>
            </li>
          ))}
        </InlineList>

        {renderCardActions ? (
          <PokemonCardActions>{renderCardActions()}</PokemonCardActions>
        ) : null}
      </PokemonCardContent>
    </CardWrapper>
  );
};

export default PokemonCard;
