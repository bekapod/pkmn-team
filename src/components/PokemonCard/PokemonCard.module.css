.content {
  padding: var(--spacing-6) 0;
  align-items: center;
}

.sprite {
  height: calc(var(--spacing-6) * 3);
}

.actions {
  margin-top: var(--spacing-6);
}
