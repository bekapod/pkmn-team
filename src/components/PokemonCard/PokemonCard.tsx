/* eslint-disable @typescript-eslint/no-empty-function */
import { FunctionComponent } from 'react';
import classNames from 'classnames';
import {
  Pokemon,
  Moves,
  TeamMemberFragmentFragment
} from '~/generated/graphql';
import { formatPokemonName, sortBySlug } from '~/lib/general';
import { CardContent, CardHeader, CardWrapper, CardHeading } from '../Card';
import { InlineList } from '../InlineList';
import { TypeTag } from '../TypeTag';
import { MoveList } from '../MoveList';

export type PokemonCardProps = {
  teamMember?: TeamMemberFragmentFragment;
  pokemon: Pokemon;
  moves?: Moves[];
  renderCardActions?: () => JSX.Element;
};

export const PokemonCard: FunctionComponent<PokemonCardProps> = ({
  teamMember,
  pokemon,
  moves = [],
  renderCardActions
}) => {
  const { pokedex_id, types, name, sprite } = pokemon;
  const actualTypes = types.map(({ type }) => type);

  return (
    <CardWrapper data-testid={`pokemon-${pokemon.pokedex_id}`}>
      <CardHeader types={actualTypes}>
        <CardHeading>{formatPokemonName(pokemon)}</CardHeading>
      </CardHeader>

      <CardContent className={classNames('py-6', 'items-center')}>
        <img
          className={classNames('h-14')}
          src={`/sprites/${sprite}`}
          alt={`${name} sprite`}
        />

        <InlineList>
          {sortBySlug(actualTypes).map(type => (
            <li
              key={`${
                teamMember ? `Member: ${teamMember.id}` : ''
              } Pokemon: ${pokedex_id}, Type: ${type.slug}`}
            >
              <TypeTag type={type.slug}>{type.name}</TypeTag>
            </li>
          ))}
        </InlineList>

        {moves.length > 0 && (
          <MoveList
            className="mt-6 w-full"
            initialMoves={moves}
            updateTeamMemberMove={() => {}}
            removeMoveFromTeamMember={() => {}}
          />
        )}

        {renderCardActions ? (
          <div className={classNames('mt-6')}>{renderCardActions()}</div>
        ) : null}
      </CardContent>
    </CardWrapper>
  );
};
