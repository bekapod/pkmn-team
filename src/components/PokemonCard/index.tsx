import React, { ReactNode, ReactNodeArray } from "react";
import styled from "styled-components/macro";
import { formatPokemonName, sortTypes } from "../../helpers/general";
import * as variables from "../../helpers/variables";
import { IPokemon, Type } from "../../types";
import { CardContent, CardHeader, CardWrapper } from "../Card";
import CardHeading from "../CardHeading";
import InlineList from "../InlineList";
import TypeTag from "../TypeTag";

interface IProps {
  memberId?: string;
  pokemon: IPokemon;
  renderCardActions?: () => ReactNode | ReactNodeArray;
}

const PokemonCardContent = styled(CardContent)`
  padding: ${variables.spacing.lg}px 0;
  align-items: center;
`;

const PokemonCardSprite = styled.img`
  height: calc(${variables.spacing.lg}px * 3);
`;

const PokemonCardActions = styled.div`
  margin-top: ${variables.spacing.lg}px;
`;

const PokemonCard = ({ memberId, pokemon, renderCardActions }: IProps) => {
  const { pokedexId, types = [], name, sprite } = pokemon;

  return (
    <CardWrapper>
      <CardHeader types={types}>
        <CardHeading>{formatPokemonName(pokemon)}</CardHeading>
      </CardHeader>

      <PokemonCardContent>
        <PokemonCardSprite src={`/sprites/${sprite}`} alt={`${name} sprite`} />

        <InlineList>
          {sortTypes(types).map((type: Type) => (
            <li
              key={`${
                memberId ? `Member: ${memberId}` : ""
              } Pokemon: ${pokedexId}, Type: ${type}`}
            >
              <TypeTag type={type}>{type}</TypeTag>
            </li>
          ))}
        </InlineList>

        {renderCardActions ? (
          <PokemonCardActions>{renderCardActions()}</PokemonCardActions>
        ) : null}
      </PokemonCardContent>
    </CardWrapper>
  );
};

export default PokemonCard;
