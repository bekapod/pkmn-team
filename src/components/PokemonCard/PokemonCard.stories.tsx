import { Meta, Story } from '@storybook/react/types-6-0';
import { pikachu } from '~/mocks/Pokemon';
import { CtaButton } from '~/components/Cta';
import { MovesProvider } from '~/hooks/useMoves';
import { PokemonCard, PokemonCardProps } from './PokemonCard';
import { explosion, flash } from '~/mocks/Pokemon_Move';

export default {
  title: 'Components/PokemonCard',
  component: PokemonCard,
  args: {
    pokemon: pikachu
  }
} as Meta<PokemonCardProps>;

export const Standard: Story<PokemonCardProps> = args => (
  <PokemonCard {...args} />
);

export const WithTeamMember: Story<
  PokemonCardProps & { updateTeamMemberMoves: any }
> = args => (
  <MovesProvider
    teamMember={args.teamMember}
    updateTeamMemberMoves={args.updateTeamMemberMoves}
  >
    <PokemonCard {...args} />
  </MovesProvider>
);
WithTeamMember.argTypes = {
  updateTeamMemberMoves: { action: 'updateTeamMemberMoves' }
};
WithTeamMember.args = {
  teamMember: {
    id: '3',
    slot: 3,
    pokemon: pikachu,
    moves: {
      total: 2,
      teamMemberMoves: [
        {
          id: 'explosion',
          slot: 1,
          move: explosion
        },
        {
          id: 'flash',
          slot: 2,
          move: flash
        }
      ]
    }
  },
  renderCardActions: () => (
    <CtaButton type="button" size="small" variant="destructive">
      Some action
    </CtaButton>
  )
};

export const WithActions: Story<PokemonCardProps> = args => (
  <PokemonCard {...args} />
);
WithActions.args = {
  renderCardActions: () => (
    <CtaButton type="button" size="small" variant="destructive">
      Some action
    </CtaButton>
  )
};
