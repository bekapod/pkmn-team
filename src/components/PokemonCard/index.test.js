// @flow
import React from "react";
import renderer from "react-test-renderer";
import PokemonCard from ".";

describe("<PokemonCard />", () => {
  it("renders without crashing", () => {
    const tree = renderer.create(
      <PokemonCard
        pokemon={{
          name: "pikachu",
          pokedexId: 25,
          sprite: "25.png",
          types: ["ELECTRIC"]
        }}
      />
    );

    expect(tree.toJSON()).toMatchSnapshot();
  });

  describe("when pokemon is part of a team (has a member id)", () => {
    it("renders without crashing", () => {
      const tree = renderer.create(
        <PokemonCard
          memberId="7"
          pokemon={{
            name: "pikachu",
            pokedexId: 25,
            sprite: "25.png",
            types: ["ELECTRIC"]
          }}
        />
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe("when a renderCardActions function is passed", () => {
    it("renders the card actions", () => {
      const tree = renderer.create(
        <PokemonCard
          pokemon={{
            name: "pikachu",
            pokedexId: 25,
            sprite: "25.png",
            types: ["ELECTRIC"]
          }}
          renderCardActions={() => (
            <div>Card actions that should be rendered</div>
          )}
        />
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe("when a pokemon with more than one type is passed", () => {
    it("renders all types", () => {
      const tree = renderer.create(
        <PokemonCard
          pokemon={{
            name: "bulbasaur",
            pokedexId: 1,
            sprite: "1.png",
            types: ["POISON", "GRASS"]
          }}
        />
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe("when a pokemon with no type is passed", () => {
    it("renders without crashing", () => {
      const tree = renderer.create(
        <PokemonCard
          pokemon={{
            name: "pikachu",
            pokedexId: 25,
            sprite: "25.png"
          }}
        />
      );

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });

  describe("when a pokemon with incomplete / invalid data is passed", () => {
    it("renders without crashing", () => {
      const tree = renderer.create(<PokemonCard pokemon={{}} />);

      expect(tree.toJSON()).toMatchSnapshot();
    });
  });
});
