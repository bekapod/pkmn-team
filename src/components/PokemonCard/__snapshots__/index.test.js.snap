// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonCard /> renders without crashing 1`] = `
<article
  className="Card__CardWrapper-sc-1prdgex-0 hFLESf"
>
  <header
    className="Card__CardHeader-sc-1prdgex-1 gOHdVf"
  >
    <h2
      className="CardHeading-esgf7h-0 hqHCZd"
    >
      #25 Pikachu
    </h2>
  </header>
  <div
    className="Card__CardContent-sc-1prdgex-2 PokemonCard__PokemonCardContent-rhhzvs-0 iHtanm"
  >
    <img
      alt="pikachu sprite"
      className="PokemonCard__PokemonCardSprite-rhhzvs-1 iixLRW"
      src="/sprites/25.png"
    />
    <ul
      className="InlineList-bwsyp4-0 bKIwJu"
    >
      <li>
        <span
          className="TypeTag-sc-19epxam-0 eoDTXf"
          type="ELECTRIC"
        >
          ELECTRIC
        </span>
      </li>
    </ul>
  </div>
</article>
`;

exports[`<PokemonCard /> when a pokemon with incomplete / invalid data is passed renders without crashing 1`] = `
<article
  className="Card__CardWrapper-sc-1prdgex-0 hFLESf"
>
  <header
    className="Card__CardHeader-sc-1prdgex-1 jVbTxc"
  >
    <h2
      className="CardHeading-esgf7h-0 hqHCZd"
    >
      #undefined 
    </h2>
  </header>
  <div
    className="Card__CardContent-sc-1prdgex-2 PokemonCard__PokemonCardContent-rhhzvs-0 iHtanm"
  >
    <img
      alt="undefined sprite"
      className="PokemonCard__PokemonCardSprite-rhhzvs-1 iixLRW"
      src="/sprites/undefined"
    />
    <ul
      className="InlineList-bwsyp4-0 bKIwJu"
    />
  </div>
</article>
`;

exports[`<PokemonCard /> when a pokemon with more than one type is passed renders all types 1`] = `
<article
  className="Card__CardWrapper-sc-1prdgex-0 hFLESf"
>
  <header
    className="Card__CardHeader-sc-1prdgex-1 dXwGGs"
  >
    <h2
      className="CardHeading-esgf7h-0 hqHCZd"
    >
      #1 Bulbasaur
    </h2>
  </header>
  <div
    className="Card__CardContent-sc-1prdgex-2 PokemonCard__PokemonCardContent-rhhzvs-0 iHtanm"
  >
    <img
      alt="bulbasaur sprite"
      className="PokemonCard__PokemonCardSprite-rhhzvs-1 iixLRW"
      src="/sprites/1.png"
    />
    <ul
      className="InlineList-bwsyp4-0 bKIwJu"
    >
      <li>
        <span
          className="TypeTag-sc-19epxam-0 dCWWOO"
          type="GRASS"
        >
          GRASS
        </span>
      </li>
      <li>
        <span
          className="TypeTag-sc-19epxam-0 daqRLF"
          type="POISON"
        >
          POISON
        </span>
      </li>
    </ul>
  </div>
</article>
`;

exports[`<PokemonCard /> when a pokemon with no type is passed renders without crashing 1`] = `
<article
  className="Card__CardWrapper-sc-1prdgex-0 hFLESf"
>
  <header
    className="Card__CardHeader-sc-1prdgex-1 jVbTxc"
  >
    <h2
      className="CardHeading-esgf7h-0 hqHCZd"
    >
      #25 Pikachu
    </h2>
  </header>
  <div
    className="Card__CardContent-sc-1prdgex-2 PokemonCard__PokemonCardContent-rhhzvs-0 iHtanm"
  >
    <img
      alt="pikachu sprite"
      className="PokemonCard__PokemonCardSprite-rhhzvs-1 iixLRW"
      src="/sprites/25.png"
    />
    <ul
      className="InlineList-bwsyp4-0 bKIwJu"
    />
  </div>
</article>
`;

exports[`<PokemonCard /> when a renderCardActions function is passed renders the card actions 1`] = `
<article
  className="Card__CardWrapper-sc-1prdgex-0 hFLESf"
>
  <header
    className="Card__CardHeader-sc-1prdgex-1 gOHdVf"
  >
    <h2
      className="CardHeading-esgf7h-0 hqHCZd"
    >
      #25 Pikachu
    </h2>
  </header>
  <div
    className="Card__CardContent-sc-1prdgex-2 PokemonCard__PokemonCardContent-rhhzvs-0 iHtanm"
  >
    <img
      alt="pikachu sprite"
      className="PokemonCard__PokemonCardSprite-rhhzvs-1 iixLRW"
      src="/sprites/25.png"
    />
    <ul
      className="InlineList-bwsyp4-0 bKIwJu"
    >
      <li>
        <span
          className="TypeTag-sc-19epxam-0 eoDTXf"
          type="ELECTRIC"
        >
          ELECTRIC
        </span>
      </li>
    </ul>
    <div
      className="PokemonCard__PokemonCardActions-rhhzvs-2 dqxORH"
    >
      <div>
        Card actions that should be rendered
      </div>
    </div>
  </div>
</article>
`;

exports[`<PokemonCard /> when pokemon is part of a team (has a member id) renders without crashing 1`] = `
<article
  className="Card__CardWrapper-sc-1prdgex-0 hFLESf"
>
  <header
    className="Card__CardHeader-sc-1prdgex-1 gOHdVf"
  >
    <h2
      className="CardHeading-esgf7h-0 hqHCZd"
    >
      #25 Pikachu
    </h2>
  </header>
  <div
    className="Card__CardContent-sc-1prdgex-2 PokemonCard__PokemonCardContent-rhhzvs-0 iHtanm"
  >
    <img
      alt="pikachu sprite"
      className="PokemonCard__PokemonCardSprite-rhhzvs-1 iixLRW"
      src="/sprites/25.png"
    />
    <ul
      className="InlineList-bwsyp4-0 bKIwJu"
    >
      <li>
        <span
          className="TypeTag-sc-19epxam-0 eoDTXf"
          type="ELECTRIC"
        >
          ELECTRIC
        </span>
      </li>
    </ul>
  </div>
</article>
`;
